### ISSUE TITLE
Measure how often layout queries are blocked by ongoing layout

### ISSUE THREAD

**Issue by jdm** [servo/servo#19797]
Servo uses a separate thread for performing layout from the thread that runs JS. When JS code triggers a synchronous layout query (such as the `getBoundingClientRectangle` API) we would like to record if the query is blocked on an ongoing layout operation in the layout thread. When invoking any of the operations that cause one of [these messages](https://github.com/servo/servo/blob/8a740aa4d1e75aad9308e3eb805ed3e15078fb59/components/script_layout_interface/message.rs#L114-L124) to be sent to the layout thread, we want to check if the layout thread is busy. We can increase a counter that we store in [ScriptThread](https://github.com/servo/servo/blob/8a740aa4d1e75aad9308e3eb805ed3e15078fb59/components/script/script_thread.rs#L393) based on the result, and display statistics about the data during [shutdown](https://github.com/servo/servo/blob/8a740aa4d1e75aad9308e3eb805ed3e15078fb59/components/script/script_thread.rs#L1905).

To determine if layout is busy, we can use an `Arc<AtomicBoolean>` value that is shared by both threads. It can be stored as a member of [Window](https://github.com/servo/servo/blob/8a740aa4d1e75aad9308e3eb805ed3e15078fb59/components/script/dom/window.rs#L160) and [LayoutThread](https://github.com/servo/servo/blob/8a740aa4d1e75aad9308e3eb805ed3e15078fb59/components/layout_thread/lib.rs#L147), and created by the script thread and sent as part of the [CreateLayoutThread](https://github.com/servo/servo/blob/fa319170ebb34afcdfc120b7c3e47fe5b1c21210/components/script_layout_interface/message.rs#L84) message.

Code:
* `components/script/script_thread.rs`
* `components/layout_thread/lib.rs`
* `components/script_layout_interface/message.rs`
* `components/script/dom/window.rs`
* `components/constellation/pipeline.rs` (the initial script and layout threads created when loading a document will need to have the AtomicBool value created for them in `UnprivilegedPipelineContent::start_all`)

Note: this is intended as a student project. Please talk to me before starting to work on it.

**Comment by jdm** [comment]
This is available for anybody who is interested in working on it.

**Comment by aishpant** [comment]
I would like to work on this.

**Comment by jdm** [comment]
@aishpant That would be great! Please ask questions if anything is unclear!

**Comment by jdm** [comment]
@aishpant Have you made any progress with this task?

**Comment by aishpant** [comment]
Sorry, I haven't had the time to look at this :-( Please assign it to someone else.

**Comment by Gaelan** [comment]
I'll give this a shot.

**Comment by jdm** [comment]
Great!

**Comment by pandusonu2** [comment]
I know this is assigned, but can I give this a try and raise a PR if I am able to get any progress?

**Comment by Gaelan** [comment]
Go for it. I'm partially done with it, but I've been busy later.

**Comment by jdm** [comment]
This is open for anybody to work on.

**Comment by jdm** [comment]
It's available for anybody.

**Comment by pylbrecht** [comment]
Alright, then I'd like to work on that!

**Comment by jdm** [comment]
It's all yours! Ask questions if anything is unclear :)

**Comment by pylbrecht** [comment]
Are there any existing tests for the code involved? Should I write tests for the new code?

**Comment by jdm** [comment]
We do not have any unit tests for the code involved, and it's not clear to me that it is feasible to do so. Manual testing for these changes is fine.

**Comment by pylbrecht** [comment]
I looked into the code and just want to verify my assumptions before I continue:

Every [QueryMsg](https://doc.servo.org/script_layout_interface/message/enum.QueryMsg.html) has a corresponding [method](https://doc.servo.org/script/dom/window/struct.Window.html#method.content_box_query) in `script/dom/window.rs`, which calls [`Window::layout_reflow()`](https://doc.servo.org/script/dom/window/struct.Window.html#method.layout_reflow) (side note: I didn't check _all_ messages). So I'm thinking of putting the check, if the layout thread is busy, in `layout_reflow()`.

[`LayoutThread::handle_request_helper()`](https://doc.servo.org/layout_thread/struct.LayoutThread.html#method.handle_request_helper) seems to handle messages coming from `ScriptThread`, so I would set/unset the busy flag in there.

Does this make sense?

**Comment by jdm** [comment]
Yes, those sound like reasonable choices!

**Comment by jdm** [comment]
Storing a `Rc<Cell<u32>>` in Window would be fine.

**Comment by pylbrecht** [comment]
So far so good.

How can I verify, that my implementation is actually working? Is there any deterministic way of creating "blocked queries"?

And in what form should statistics be displayed during shutdown of the script thread?

**Comment by jdm** [comment]
The easiest way to trigger blocking queries is to have JS that modifies the layout of the page in some fashion (changing CSS; mutating the DOM by adding/removing a node; etc.) then immediately performing a layout query on a node that is affected by the changes (like getBoundingClientRectangle). We could start with reporting how many queries were blocked out of the total number of queries that were initiated for a given ScriptThread.

**Comment by pylbrecht** [comment]
Unfortunately I was not able to verify [my code](https://github.com/pylbrecht/servo/tree/measure.blocked.layout.queries), yet.

I'm trying to trigger a blocked query with this `test.html`:
```html
<!DOCTYPE html>
<html>
  <head>
    <title>Test</title>
  </head>
  <body>
    <div id="test">
      test
    </div>
    <script>
      document.getElementById("test").style.color = "blue";
      document.getElementById("test").getBoundingClientRect();
    </script>
  </body>
</html>
```
I'm running Servo with `RUST_LOG="script::script_thread"` to check for the statistics printed via `debug!()`.
I tried placing the reporting code in:
- `Drop` implementation for `ScriptThread`
- immediately after the `script_thread.start()` call in `ScriptThreadFactory::create()`
- in `ScriptThread::handle_exit_script_thread_msg()`

None of which produced the desired output.
The last lines always seem look like this (I expected to find at least a "Exiting script thread" message, printed in `handle_exit_script_thread()`):
```
05.368657:  DEBUG 2018-08-20T21:22:40Z: script::script_thread: Got event.
05.368830:  DEBUG 2018-08-20T21:22:40Z: script::script_thread: Processing events.
05.368892:  DEBUG 2018-08-20T21:22:40Z: script::script_thread: Processing event FromConstellation(ConstellationControlMsg::SendEvent).
05.369359:  DEBUG 2018-08-20T21:22:40Z: script::script_thread: Issuing batched reflows.
05.369471:  DEBUG 2018-08-20T21:22:40Z: script::script_thread: Running script thread.
05.369526:  DEBUG 2018-08-20T21:22:40Z: script::script_thread: Waiting for event.
06.242025:  DEBUG 2018-08-20T21:22:41Z: script::script_thread: Got event.
06.242118:  DEBUG 2018-08-20T21:22:41Z: script::script_thread: Processing events.
06.242163:  DEBUG 2018-08-20T21:22:41Z: script::script_thread: Processing event FromConstellation(ConstellationControlMsg::ExitPipeline).
06.242199:  DEBUG 2018-08-20T21:22:41Z: script::script_thread: Exiting pipeline (0,1).
06.242373:  DEBUG 2018-08-20T21:22:41Z: script::script_thread: preparing to shut down layout for page (0,1)
06.242544:  DEBUG 2018-08-20T21:22:41Z: script::script_thread: shutting down layout for page (0,1)
06.246516:  DEBUG 2018-08-20T21:22:41Z: script::script_thread: tracing fields of ScriptThread
```

**Comment by pylbrecht** [comment]
I tried this on the latest master hoping to see different results, but got the same debug messages. I don't understand how the involved threads are terminated, meaning by whom and in what order, because I don't get any more debug messages coming from the script thread than those in my last comment.
Generally speaking I don't know where to look.

Somehow I don't trust the debug output, because it abruptly stops as soon as I terminate Servo. Is there a better way of running a manual test than the following?
1. `RUST_LOG="debug" ./mach run -d`
2. Terminate Servo by clicking the "x" in the window title bar
3. Look at the generated log

**Comment by pylbrecht** [comment]
I had long forgotten this issue, but I remember being stuck at some point.

@jdm, do you still consider this an approachable issue for a servo starter like me or would another one be better suited? The issue looks kind of abandoned at the moment, so I'm unsure if it's still a good fit.

**Comment by jdm** [comment]
@pylbrecht Oh dear. I remember adding your previous question to my list of things to investigate; I apologize for never returning to it. You're right that the script thread never receives the ExitScriptThread message; the message is definitely sent, but the main thread appears to terminate before the message is received by the script thread.

**Comment by jdm** [comment]
Ok, I have a new strategy to recommend (and I do believe it's still appropriate for a starter issue!):
* add a new field to Profiler which is a map from string (url) to u32
* add a new variant to ProfilerMsg which takes a string
* every time a layout query is blocked (per previous instructions), send this new message to the profiler
* when the profiler receives this new message, increase the appropriate counter by 1
* in `Profiler::print_buckets` also print the new layout blocked counters

code: `components/profiler_traits/time.rs`, `components/profiler/time.rs`

**Comment by pylbrecht** [comment]
Great, then I'll dive right back into it! I'll keep you up to date on my progress.

**Comment by pylbrecht** [comment]
So I added all neccessary fields and my code compiles without any errors.
What's still missing is the communication with the profiler, which I just skimmed over so far.

Though I am not so sure about my manual testing approach from [my previous comment](https://github.com/servo/servo/issues/19797#issuecomment-414471651). I added a `panic!()` to the "layout thread is busy" check, but it was not triggered. So either my code is wrong or the test is insufficient.
